cogs5e/models/homebrew/bestiary.py

```diff
@@ -1,56 +1,129 @@
+                sha256_hash.update(name.encode() + desc.encode())
+
+        # try and find a bestiary by looking up upstream|hash
+        # if it exists, return it
+        # otherwise commit a new one to the db and return that
+        sha256 = sha256_hash.hexdigest()
+        log.debug(f"Bestiary hash: {sha256}")
+        existing_bestiary = await ctx.bot.mdb.bestiaries.find_one({"upstream": url, "sha256": sha256})
+        if existing_bestiary:
+            log.info("This bestiary already exists, subscribing")
+            existing_bestiary = Bestiary.from_dict(existing_bestiary)
+            await existing_bestiary.subscribe(ctx)
+            return existing_bestiary
+
+        parsed_creatures = [Monster.from_critterdb(c) for c in creatures]
+        b = cls(None, sha256, url, [], [], [], name, parsed_creatures, desc)
+        await b.write_to_db(ctx)
+        return b
 
     @classmethod
     async def from_ctx(cls, ctx):
-        active_bestiary = await ctx.bot.mdb.bestiaries.find_one({"owner": str(ctx.author.id), "active": True})
+        active_bestiary = await ctx.bot.mdb.bestiaries.find_one({"active": str(ctx.author.id)},
+                                                                projection={"monsters": False})
         if active_bestiary is None:
             raise NoActiveBrew()
-        return cls.from_raw(active_bestiary['critterdb_id'], active_bestiary)
-
-    def to_dict(self):
-        return {'monsters': [m.to_dict() for m in self.monsters], 'name': self.name, 'critterdb_id': self.id,
-                'desc': self.desc}
-
-    async def commit(self, ctx):
-        """Writes a bestiary object to the database, under the contextual author. Returns self."""
-        data = {"$set": self.to_dict(), "$setOnInsert": {"owner": str(ctx.author.id), "server_active": []}}
-
-        await ctx.bot.mdb.bestiaries.update_one(
-            {"owner": str(ctx.author.id), "critterdb_id": self.id},
-            data,
-            True
-        )
+        return cls.from_dict(active_bestiary)
+
+    async def load_monsters(self, ctx):
+        if not self._monsters:
+            bestiary = await ctx.bot.mdb.bestiaries.find_one({"_id": self.id}, projection=['monsters'])
+            self._monsters = [Monster.from_bestiary(m) for m in bestiary['monsters']]
+        return self._monsters
+
+    @property
+    def monsters(self):
+        if not self._monsters:
+            raise AttributeError("load_monsters() must be called before accessing bestiary monsters.")
+        return self._monsters
+
+    async def write_to_db(self, ctx):
+        """Writes a bestiary object to the database. Returns self."""
+        assert self._monsters is not None
+        data = {
+            "sha256": self.sha256, "upstream": self.upstream, "subscribers": [str(ctx.author.id)], "active": [],
+            "server_active": [], "name": self.name, "desc": self.desc, "monsters": [m.to_dict() for m in self._monsters]
+        }
```

Comment:
I find this difficult to parse, one key:value pair per line is much easier for me at least.